### PR TITLE
Issue 88 - Add IdentifiableType filter

### LINKED ISSUE #88: Filter substations and voltage levels according to an Identifiable type
### Describe the current behavior

Theviewer left panel displays the list of susbtations and voltage levels of the studied network.
image
In the current powsybl-dev-tools version, it is only possible to filter substations and voltage levels by name:
image

### Describe the expected behavior

We would like to be able to filter substations and their voltage levels according to the type of Identifiable they contain.
Users can only select one type of Identifiable at a time.

### Describe the motivation

When working on a specific feature or studying a specific aspect of the network, e.g. tie lines, it is useful to easily access and display the voltage levels that contain them.

### Extra Information

_No response_

### PR DESCRIPTION
**Pull request by tadam50**
**Please check if the PR fulfills these requirements**

- [x] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem?**
'Fixes #88'

### CODE REVIEW

**Comment by tadam50** [comment]
I suggest to use IdentifiableType enumearation as filter on connectable components.
If we use this enumeraration, we could use NETWORK value to keep all components (no filtering use case).
But i've no use case in mind to use SUBSTATION and VOLTAGE_LEVEL items.
Are these items need to be removed from list in filtering widget ?

See following animation for example :
issue-88

File: diagram-viewer/src/main/java/com/powsybl/diagram/viewer/MainViewController.java [hunk 1]
```diff
@@ -302,6 +309,20 @@ private void initSubstationsTree(Network network) {
         vlTree.setShowRoot(true);
     }
 
+    private boolean testComponent(IdentifiableType type, Identifiable<?> identifiable) {
+        boolean result = false;
+        if (type == IdentifiableType.NETWORK || type == IdentifiableType.SUBSTATION || type == IdentifiableType.VOLTAGE_LEVEL) {
```
Comment (flo-dup) [inline diagram-viewer/src/main/java/com/powsybl/diagram/viewer/MainViewController.java:314]: ok for network and voltage level, but I would expect that choosing substation removes the voltage levels which have no substation (directly under network)

File: diagram-viewer/src/main/java/com/powsybl/diagram/viewer/MainViewController.java [hunk 2]
```diff
@@ -302,6 +309,20 @@ private void initSubstationsTree(Network network) {
...
+            result = true;
+        } else {
+            if (identifiable instanceof Substation s) {
+                result = s.getVoltageLevelStream().anyMatch(v -> testComponent(type, v));
+            } else if (identifiable instanceof VoltageLevel v) {
+                result = v.getConnectableStream().anyMatch(c -> c.getType() == type);
```
Comment (flo-dup) [inline diagram-viewer/src/main/java/com/powsybl/diagram/viewer/MainViewController.java:320]: Does not work for switches nor buses. One would expect that selecting `BUS` item would remove the voltage levels with no bus in the bus view?

File: diagram-viewer/src/main/resources/mainView.fxml [hunk 3]
```diff
@@ -28,7 +26,10 @@
                     <CheckBox fx:id="hideSubstations" onMouseClicked="#initSubstationsTree" text="Hide substations" GridPane.rowIndex="2"/>
                     <CheckBox fx:id="hideVoltageLevels" onMouseClicked="#initSubstationsTree" text="Hide voltage levels" GridPane.rowIndex="3"/>
                     <Label text="Filter:" GridPane.rowIndex="4"/>
-                    <TextField fx:id="filterField" onKeyTyped="#initSubstationsTree" minWidth="40" GridPane.rowIndex="5"/>
+                    <HBox GridPane.rowIndex="5" spacing="5">
+                        <TextField fx:id="filterField" onKeyTyped="#initSubstationsTree" minWidth="40"/>
+                        <EnumChoiceBox enumType="com.powsybl.iidm.network.IdentifiableType" fx:id="componentTypeFilterChoice" initialValue="All" onAction="#initSubstationsTree"/>
```
Comment (flo-dup) [inline diagram-viewer/src/main/resources/mainView.fxml:31]: It's confusing to put it here: one would expect the filter string `filterField` to apply only on the type of equipment chosen. We need a title above the choice box to make it clearer.

**Review by flo-dup — CHANGES_REQUESTED** [review]
Nice feature! it will save us time each time we develop a feature on a specific equipment :+1:

File: diagram-viewer/src/main/resources/mainView.fxml (hunk 3 above)
Comment (tadam50) [reply diagram-viewer/src/main/resources/mainView.fxml:31]: This is an other solution:<BR>
image

**Review by So-Fras — APPROVED** [review]

**Comment by So-Fras** [comment]
Adding the phase-shift transformers to the list is very useful but I think that (and the refactoring that ensues) should be done in another PR.

File: diagram-viewer/src/main/java/com/powsybl/diagram/viewer/MainViewController.java [hunk 4]
```diff
@@ -380,6 +456,23 @@ private void initSubstationsTree(Network network) {
         vlTree.setShowRoot(true);
     }
 
+    private boolean testComponent(ComponentFilterType type, Identifiable<?> identifiable) {
```
Comment (flo-dup) [inline diagram-viewer/src/main/java/com/powsybl/diagram/viewer/MainViewController.java:459]: We only call this for voltage levels and substations, so we can be more precise than the identifiable class. The name can also be more explicit. And the method can be static.
```suggestion
    private static boolean containsComponentType(ComponentFilterType type, Container<?> container) {
```

File: diagram-viewer/src/main/java/com/powsybl/diagram/viewer/MainViewController.java [hunk 5]
```diff
@@ -43,6 +43,75 @@
  */
 public class MainViewController {
 
+    public enum ComponentFilterType {
+        ALL,
+        HVDC_LINE,
+        BUS,
```
Comment (flo-dup) [inline diagram-viewer/src/main/java/com/powsybl/diagram/viewer/MainViewController.java:49]: I think we can remove this enum, I was just trying how to make sense for the IdentifiableType BUS, but I don't think it's of interest to have that filtering. The opposite would be interesting though (no buses).

File: diagram-viewer/src/main/java/com/powsybl/diagram/viewer/MainViewController.java [hunk 6]
```diff
@@ -43,6 +43,75 @@
...
+                case LOAD -> IdentifiableType.LOAD;
+                case SHUNT_COMPENSATOR -> IdentifiableType.SHUNT_COMPENSATOR;
+                case DANGLING_LINE -> IdentifiableType.DANGLING_LINE;
+                case STATIC_VAR_COMPENSATOR -> IdentifiableType.STATIC_VAR_COMPENSATOR;
+                case VSC_CONVERTER_STATION, LCC_CONVERTER_STATION -> IdentifiableType.HVDC_CONVERTER_STATION;
+                case GROUND -> IdentifiableType.GROUND;
+            };
+        }
+
+        ComponentFilterType() { /* compiled code */ }
```
Comment (flo-dup) [inline diagram-viewer/src/main/java/com/powsybl/diagram/viewer/MainViewController.java:88]: can't we remove this?

File: diagram-viewer/src/main/java/com/powsybl/diagram/viewer/MainViewController.java [hunk 7]
```diff
@@ -43,6 +43,75 @@
...
+                case TIE_LINE -> IdentifiableType.TIE_LINE;
+                case TWO_WINDINGS_TRANSFORMER -> IdentifiableType.TWO_WINDINGS_TRANSFORMER;
+                case THREE_WINDINGS_TRANSFORMER -> IdentifiableType.THREE_WINDINGS_TRANSFORMER;
+                case GENERATOR -> IdentifiableType.GENERATOR;
+                case BATTERY -> IdentifiableType.BATTERY;
+                case LOAD -> IdentifiableType.LOAD;
+                case SHUNT_COMPENSATOR -> IdentifiableType.SHUNT_COMPENSATOR;
+                case DANGLING_LINE -> IdentifiableType.DANGLING_LINE;
+                case STATIC_VAR_COMPENSATOR -> IdentifiableType.STATIC_VAR_COMPENSATOR;
+                case VSC_CONVERTER_STATION, LCC_CONVERTER_STATION -> IdentifiableType.HVDC_CONVERTER_STATION;
```
Comment (flo-dup) [inline diagram-viewer/src/main/java/com/powsybl/diagram/viewer/MainViewController.java:83]: I was wondering how you make any difference between them then! but then I saw you get it from VscCount / LccCount method :+1:

File: diagram-viewer/src/main/java/com/powsybl/diagram/viewer/MainViewController.java [hunk 8]
```diff
@@ -43,6 +43,75 @@
...
+            return switch (this) {
+                case ALL -> "All";
+                case HVDC_LINE -> "HVDC line";
+                case BUS -> "Bus";
+                case SWITCH -> "Switch";
+                case BUSBAR_SECTION -> "Busbar section";
+                case LINE -> "Line";
+                case TIE_LINE -> "Tie line";
+                case TWO_WINDINGS_TRANSFORMER -> " Two winding transformer";
+                case THREE_WINDINGS_TRANSFORMER -> "Three winding transformer";
```
Comment (flo-dup) [inline diagram-viewer/src/main/java/com/powsybl/diagram/viewer/MainViewController.java:101]: nice, you didn't put the 's', which is an old powsybl mistake! but you should put the dash though
```suggestion
                case TWO_WINDINGS_TRANSFORMER -> " Two-winding transformer";
                case THREE_WINDINGS_TRANSFORMER -> "Three-winding transformer";
```

**Review by flo-dup — CHANGES_REQUESTED** [review]

**Review by flo-dup — APPROVED** [review]

**Pull request merged**